src/ir/bits.h

```diff
@@ -438,10 +439,55 @@ Index getMaxBits(Expression* curr,
 }
 
 // As getMaxBits, but returns the minimum amount of bits.
-inline Index getMinBits(Expression* curr) {
+template<typename LocalInfoProvider = DummyLocalInfoProvider>
+inline Index getMinBits(Expression* curr,
+                        LocalInfoProvider* localInfoProvider = nullptr) {
   if (auto* c = curr->dynCast<Const>()) {
     // Constants are simple: the min and max are identical.
     return getMaxBits(c);
+  } else if (auto* binary = curr->dynCast<Binary>()) {
+    switch (binary->op) {
+      case AndInt32:
+      case AndInt64: {
+        return std::max(getMinBits(binary->left, localInfoProvider),
+                        getMinBits(binary->right, localInfoProvider));
+      }
+      case OrInt32:
+      case OrInt64:
+      case XorInt32:
+      case XorInt64: {
+        return std::min(getMinBits(binary->left, localInfoProvider),
+                        getMinBits(binary->right, localInfoProvider));
+      }
+      case ShlInt32:
+      case ShlInt64: {
+        if (auto* shift = binary->right->dynCast<Const>()) {
+          Index maxBits = shift->type.getByteSize() * 8;
+          Index minBits = getMinBits(binary->left, localInfoProvider);
+          return std::min(maxBits, minBits + Bits::getEffectiveShifts(shift));
+        }
+        break;
+      }
+      case ShrUInt32:
+      case ShrUInt64:
+      case ShrSInt32:
+      case ShrSInt64: {
+        if (auto* shift = binary->right->dynCast<Const>()) {
+          Index minBits = getMinBits(binary->left, localInfoProvider);
+          return std::max(Index(0), minBits - Bits::getEffectiveShifts(shift));
```

Comment:
This doesn't look right. Max of an unsigned 0 and another value does nothing. I think the values in the max should both be signed.